gateway/src/main/kotlin/handler/HandshakeHandler.kt

```diff
@@ -19,31 +19,37 @@ internal class HandshakeHandler(
 
     lateinit var configuration: GatewayConfiguration
 
-    private val session: AtomicRef<String?> = atomic(null)
+    // see https://discord.com/developers/docs/topics/gateway#resuming
+    private class ResumeContext(val sessionId: String, val resumeUrl: Url)
 
-    private val identify
-        get() = configuration.identify
+    private val resumeContext = atomic<ResumeContext?>(initial = null)
+    val gatewayUrl get() = resumeContext.value?.resumeUrl ?: initialUrl
 
-    private val resume
-        get() = Resume(configuration.token, session.value!!, sequence.value ?: 0)
-
-    private val sessionStart get() = session.value == null
-
-    override fun start() {
-        on<Ready> { event ->
-            session.update { event.data.sessionId }
+    private val resumeOrIdentify
+        get() = when (val sessionId = resumeContext.value?.sessionId) {
+            null -> configuration.identify
+            else -> Resume(configuration.token, sessionId, sequence.value ?: 0)
         }
 
+    override fun start() {
         on<Hello> {
-            reconnectRetry.reset() //connected and read without problems, resetting retry counter
+            reconnectRetry.reset() // connected and read without problems, resetting retry counter
             identifyRateLimiter.consume {
-                if (sessionStart) send(identify)
-                else send(resume)
+                send(resumeOrIdentify)
             }
```

Comment:
Do we also need to rate-limit resumes?

[This](https://discord.com/developers/docs/topics/gateway#rate-limiting) suggests we do not:
> Clients also have a limit of [concurrent](https://discord.com/developers/docs/topics/gateway#session-start-limit-object) [Identify](https://discord.com/developers/docs/topics/gateway#identify) requests allowed per 5 seconds. If you hit this limit, the Gateway will respond with an [Opcode 9 Invalid Session](https://discord.com/developers/docs/topics/gateway#invalid-session).

And there is also [this](https://discord.com/developers/docs/topics/gateway#identifying):
> Clients are limited to 1000 `IDENTIFY` calls to the websocket in a 24-hour period. This limit is global and across all shards, but does not include `RESUME` calls.

Reply:
out of scope for this PR

Reply:
In fact this is solved by #693